zealot/fetcher/fetcher.ts

```diff
@@ -1,72 +1,61 @@
 import {url} from "../util/utils"
 import {parseContentType} from "./contentType"
-import {Enhancer, ZealotUploadPayload, ZResponse} from "../types"
+import {Enhancer, ZResponse} from "../types"
 import {createIterator} from "./iterator"
 import {createStream} from "./stream"
 import {createError} from "../util/error"
-import {createPushableIterator} from "./pushable-iterator"
-import {parseLines} from "../ndjson/lines"
+import {EventSourcePolyfill} from "event-source-polyfill"
+import nodeFetch from "node-fetch"
 
 export type FetchArgs = {
   path: string
-  method: string
-  body?: string | FormData | ReadableStream
+  method?: string
+  body?: string | FormData | ReadableStream | NodeJS.ReadableStream
   headers?: Headers
   enhancers?: Enhancer[]
   signal?: AbortSignal
+  useNodeFetch?: boolean
 }
 
 export function createFetcher(host: string) {
   return {
     async promise(args: FetchArgs) {
-      const {path, method, body, signal, headers} = args
-      const resp = await fetch(url(host, path), {method, body, signal, headers})
+      const {path, method, body, signal, headers, useNodeFetch} = args
+      const switchFetch = useNodeFetch ? nodeFetch : fetch
+      const resp = await switchFetch(url(host, path), {
+        method,
+        body,
+        signal,
+        headers
+      })
       const content = await parseContentType(resp)
       return resp.ok ? content : Promise.reject(createError(content))
     },
     async stream(args: FetchArgs): Promise<ZResponse> {
       const {path, method, body, signal, headers} = args
-      const resp = await fetch(url(host, path), {method, body, signal, headers})
+      const resp = await fetch(url(host, path), {
+        method,
+        body: body as string | FormData | ReadableStream,
+        signal,
+        headers
+      })
       if (!resp.ok) {
         const content = await parseContentType(resp)
         return Promise.reject(createError(content))
       }
       const iterator = createIterator(resp, args)
       return createStream(iterator, resp)
     },
-    async upload(args: FetchArgs): Promise<ZResponse> {
```

Comment:
Can this be removed now? How do the progress updates happen now? The events endpoint?

Reply:
Yeah we haven't been using this since I think we switched to use the load endpoint. There, we use nodeFetch and attach the request body as a stream then listen to the backend for progress updates.

Reply:
Sweet, thanks for the explanation.